File: khmer-bun/src/dictionary.ts
import * as Constants from './constants';

// Coeng characters for variant generation
const COENG_TA = '\u17D2\u178F';
const COENG_DA = '\u17D2\u178D';
const COENG_RO = '\u17D2\u179A';

// Pattern for Coeng Ro swaps
const PATTERN_RO_OTHER = /(\u17D2\u179A)(\u17D2[^\u179A])/g;
const PATTERN_OTHER_RO = /(\u17D2[^\u179A])(\u17D2\u179A)/g;

// Valid single-character words
const VALID_SINGLE_WORDS = new Set([
  'ក', 'ខ', 'គ', 'ង', 'ច', 'ឆ', 'ញ', 'ដ', 'ត', 'ទ', 'ព', 'រ', 'ល', 'ស', 'ឡ',
  'ឬ', 'ឮ', 'ឪ', 'ឯ', 'ឱ', 'ឦ', 'ឧ', 'ឳ'
]);

// Trie node for efficient prefix lookup
class TrieNode {
  children: Map<number, TrieNode> = new Map();
  cost: number = -1;
  isWord: boolean = false;
}

export class Dictionary {
  private root: TrieNode = new TrieNode();
  private words: Set<string> = new Set();
  public maxWordLength: number = 0;
  public unknownCost: number = 10.0;

  /**
   * Generates interchangeable variants for a word
   */
  private generateVariants(word: string): Set<string> {
    const variants = new Set<string>();

    // 1. Coeng Ta <-> Coeng Da
    if (word.includes(COENG_TA)) {
      variants.add(word.replaceAll(COENG_TA, COENG_DA));
    }
    if (word.includes(COENG_DA)) {
      variants.add(word.replaceAll(COENG_DA, COENG_TA));
    }

    // 2. Coeng Ro Ordering
    const baseSet = new Set([word, ...variants]);

    for (const w of baseSet) {
      // Apply Swap 1: Ro -> Other ==> Other -> Ro
      PATTERN_RO_OTHER.lastIndex = 0;
      if (PATTERN_RO_OTHER.test(w)) {
        PATTERN_RO_OTHER.lastIndex = 0;
        variants.add(w.replace(PATTERN_RO_OTHER, '$2$1'));
      }

      // Apply Swap 2: Other -> Ro ==> Ro -> Other
      PATTERN_OTHER_RO.lastIndex = 0;
      if (PATTERN_OTHER_RO.test(w)) {
        PATTERN_OTHER_RO.lastIndex = 0;
        variants.add(w.replace(PATTERN_OTHER_RO, '$2$1'));
      }
    }

    return variants;
  }

  /**
   * Load dictionary from file content
   */
  loadDictionary(content: string): void {
    const lines = content.split(/\r?\n/);

    for (const line of lines) {
      const w = line.trim();
      if (!w) continue;

      // Filter single-character words that are NOT valid single consonants
      if (w.length === 1 && !VALID_SINGLE_WORDS.has(w)) continue;

      this.words.add(w);

      // Generate and add variants
      const variants = this.generateVariants(w);
      for (const variant of variants) {
        if (variant !== w) {
          this.words.add(variant);
        }
      }
    }

    // Post-filter: Remove compound ORs
    this.postFilterCompoundOrs();

    // Remove standalone ៗ
    this.words.delete('ៗ');
  }

  /**
   * Post-filter compound OR words
   */
  private postFilterCompoundOrs(): void {
    const wordsToRemove = new Set<string>();

    for (const word of this.words) {
      if (word.includes('ឬ') && word.length > 1) {
        if (word.startsWith('ឬ')) {
          const suffix = word.slice(1);
          if (this.words.has(suffix)) {
            wordsToRemove.add(word);
          }
        } else if (word.endsWith('ឬ')) {
          const prefix = word.slice(0, -1);
          if (this.words.has(prefix)) {
            wordsToRemove.add(word);
          }
        } else {
          const parts = word.split('ឬ');
          if (parts.every(p => this.words.has(p) || p === '')) {
            wordsToRemove.add(word);
          }
        }
      }

      if (word.includes('ៗ')) {
        wordsToRemove.add(word);
      }

      if (word.startsWith('\u17D2')) {
        wordsToRemove.add(word);
      }
    }

    for (const w of wordsToRemove) {
      this.words.delete(w);
    }
  }

  /**
   * Load frequencies and build Trie
   */
  loadFrequencies(freqData: Record<string, number>): void {
    const MIN_FREQ_FLOOR = 5.0;

    // Build effective counts with variants
    const effectiveCounts = new Map<string, number>();
    let totalTokens = 0;

    for (const [word, count] of Object.entries(freqData)) {
      const eff = Math.max(count, MIN_FREQ_FLOOR);
      effectiveCounts.set(word, eff);

      // Add variants with SAME frequency
      const variants = this.generateVariants(word);
      for (const v of variants) {
        if (!effectiveCounts.has(v)) {
          effectiveCounts.set(v, eff);
        }
      }

      // Total tokens from primary words only
      totalTokens += eff;
    }

    if (totalTokens <= 0) totalTokens = 1;

    // Calculate default cost for words without frequency data
    const defaultCost = -Math.log10(MIN_FREQ_FLOOR / totalTokens);

    // Build Trie from words
    for (const word of this.words) {
      let cost: number;
      if (effectiveCounts.has(word)) {
        const count = effectiveCounts.get(word)!;
        cost = -Math.log10(count / totalTokens);
      } else {
        cost = defaultCost;
      }

      this.addWord(word, cost);
    }
  }

  /**
   * Add a word to the Trie
   */
  addWord(word: string, cost: number): void {
    let node = this.root;
    for (let i = 0; i < word.length; i++) {
      const c = word.charCodeAt(i);
      let child = node.children.get(c);
      if (!child) {
        child = new TrieNode();
        node.children.set(c, child);
      }
      node = child;
    }
    node.isWord = true;
    node.cost = cost;

    if (word.length > this.maxWordLength) {
      this.maxWordLength = word.length;
    }
  }

  /**
   * Lookup word cost by range (no substring allocation)
   */
  lookupRange(text: string, start: number, end: number): number {
    let node = this.root;
    for (let i = start; i < end; i++) {
      const c = text.charCodeAt(i);
      const child = node.children.get(c);
      if (!child) return -1;
      node = child;
    }
    return node.isWord ? node.cost : -1;
  }

  /**
   * Check if word exists in dictionary
   */
  hasWord(word: string): boolean {
    return this.words.has(word);
  }

  get size(): number {
    return this.words.size;
  }
}

File: khmer-java/src/main/java/khmer/Dictionary.java
package khmer;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;

/**
 * Dictionary and frequency model for Khmer word segmentation.
 * Optimized with Trie for fast codepoint-based lookups.
 */
public class Dictionary {

    private final Set<String> words;
    private final Map<String, Float> wordCosts;

    // Trie for fast codepoint-based lookup
    private final TrieNode trie;

    public int maxWordLength; // in codepoints
    public float defaultCost;
    public float unknownCost;

    private static final float MIN_FREQ_FLOOR = 5.0f;

    // Constants for variant generation
    private static final String COENG_TA = "\u17D2\u178F";
    private static final String COENG_DA = "\u17D2\u178D";

    public Dictionary() {
        this.words = new HashSet<>();
        this.wordCosts = new HashMap<>();
        this.trie = new TrieNode();
        this.maxWordLength = 0;
        this.defaultCost = 10.0f;
        this.unknownCost = 20.0f;
    }

    public void load(String dictionaryPath, String frequencyPath) throws IOException {
        loadDictionary(dictionaryPath);
        loadFrequencies(frequencyPath);
        buildTrie();
    }

    private void loadDictionary(String path) throws IOException {
        Set<String> validSingleWords = Constants.VALID_SINGLE_WORDS
            .stream()
            .map(String::valueOf)
            .collect(java.util.stream.Collectors.toSet());

        try (BufferedReader reader = Files.newBufferedReader(Path.of(path), StandardCharsets.UTF_8)) {
            String line;
            while ((line = reader.readLine()) != null) {
                String word = line.trim();
                if (word.isEmpty()) continue;

                // Filter invalid single-char words (use codepoint count)
                int cpCount = word.codePointCount(0, word.length());
                if (cpCount == 1 && !validSingleWords.contains(word)) {
                    continue;
                }

                addWordWithVariants(word);
            }
        }

        // Post-process: remove compound words with OR, words with repetition mark, words starting with Coeng
        Set<String> toRemove = new HashSet<>();
        for (String word : words) {
            // Contains ឬ (OR)
            if (word.contains("\u17AC") && word.codePointCount(0, word.length()) > 1) {
                if (word.startsWith("\u17AC")) {
                    String suffix = word.substring(1);
                    if (words.contains(suffix)) toRemove.add(word);
                } else if (word.endsWith("\u17AC")) {
                    String prefix = word.substring(0, word.length() - 1);
                    if (words.contains(prefix)) toRemove.add(word);
                } else {
                    String[] parts = word.split("\u17AC");
                    boolean allValid = true;
                    for (String p : parts) {
                        if (!p.isEmpty() && !words.contains(p)) {
                            allValid = false;
                            break;
                        }
                    }
                    if (allValid) toRemove.add(word);
                }
            }

            // Contains repetition mark
            if (word.contains("\u17D7")) {
                toRemove.add(word);
            }

            // Starts with Coeng
            if (word.startsWith("\u17D2")) {
                toRemove.add(word);
            }
        }

        words.removeAll(toRemove);
        words.remove("\u17D7"); // Remove standalone repetition mark

        // Recalculate max word length in codepoints
        maxWordLength = 0;
        for (String word : words) {
            int cpCount = word.codePointCount(0, word.length());
            if (cpCount > maxWordLength) {
                maxWordLength = cpCount;
            }
        }

        System.out.println("Loaded " + words.size() + " words. Max length: " + maxWordLength);
    }

    private void addWordWithVariants(String word) {
        words.add(word);
        int cpCount = word.codePointCount(0, word.length());
        if (cpCount > maxWordLength) {
            maxWordLength = cpCount;
        }

        Set<String> variants = generateVariants(word);
        for (String v : variants) {
            words.add(v);
            int vCpCount = v.codePointCount(0, v.length());
            if (vCpCount > maxWordLength) {
                maxWordLength = vCpCount;
            }
        }
    }

    private Set<String> generateVariants(String word) {
        Set<String> variants = new HashSet<>();

        // 1. Coeng Ta <-> Coeng Da swap
        if (word.contains(COENG_TA)) {
            variants.add(word.replace(COENG_TA, COENG_DA));
        }
        if (word.contains(COENG_DA)) {
            variants.add(word.replace(COENG_DA, COENG_TA));
        }

        // 2. Coeng Ro ordering swaps (without regex for performance)
        Set<String> baseSet = new HashSet<>();
        baseSet.add(word);
        baseSet.addAll(variants);

        for (String w : baseSet) {
            String swapped = swapCoengRoOrder(w);
            if (!swapped.equals(w)) {
                variants.add(swapped);
            }
        }

        return variants;
    }

    /**
     * Swap Coeng+Ro with adjacent Coeng+X patterns (no regex).
     */
    private String swapCoengRoOrder(String word) {
        int[] cps = word.codePoints().toArray();
        int n = cps.length;
        if (n < 4) return word;

        int[] result = new int[n];
        int ri = 0;
        int i = 0;
        boolean changed = false;

        while (i < n) {
            // Look for pattern: Coeng + Ro + Coeng + X
            if (i + 3 < n &&
                cps[i] == 0x17D2 && cps[i+1] == 0x179A &&
                cps[i+2] == 0x17D2 && cps[i+3] != 0x179A) {
                result[ri++] = cps[i+2];
                result[ri++] = cps[i+3];
                result[ri++] = cps[i];
                result[ri++] = cps[i+1];
                i += 4;
                changed = true;
                continue;
            }
            // Look for pattern: Coeng + X + Coeng + Ro
            if (i + 3 < n &&
                cps[i] == 0x17D2 && cps[i+1] != 0x179A &&
                cps[i+2] == 0x17D2 && cps[i+3] == 0x179A) {
                result[ri++] = cps[i+2];
                result[ri++] = cps[i+3];
                result[ri++] = cps[i];
                result[ri++] = cps[i+1];
                i += 4;
                changed = true;
                continue;
            }
            result[ri++] = cps[i++];
        }

        if (changed) {
            return new String(result, 0, ri);
        }
        return word;
    }

    private void loadFrequencies(String path) throws IOException {
        Path freqPath = Path.of(path);
        if (!Files.exists(freqPath)) {
            System.out.println("Frequency file not found at " + path + ". Using default costs.");
            return;
        }

        String json = Files.readString(freqPath, StandardCharsets.UTF_8);
        Map<String, Double> data = parseJsonMap(json);

        Map<String, Float> effectiveCounts = new HashMap<>(data.size());
        float totalTokens = 0;

        for (Map.Entry<String, Double> entry : data.entrySet()) {
            String word = entry.getKey();
            float count = entry.getValue().floatValue();
            float eff = Math.max(count, MIN_FREQ_FLOOR);
            effectiveCounts.put(word, eff);

            // Add variants with same frequency
            Set<String> variants = generateVariants(word);
            for (String v : variants) {
                effectiveCounts.putIfAbsent(v, eff);
            }

            totalTokens += eff;
        }

        if (totalTokens > 0) {
            float minProb = MIN_FREQ_FLOOR / totalTokens;
            this.defaultCost = (float) -Math.log10(minProb);
            this.unknownCost = this.defaultCost + 5.0f;

            for (Map.Entry<String, Float> entry : effectiveCounts.entrySet()) {
                float prob = entry.getValue() / totalTokens;
                if (prob > 0) {
                    wordCosts.put(entry.getKey(), (float) -Math.log10(prob));
                }
            }
        }

        System.out.println("Loaded frequencies for " + wordCosts.size() + " words.");
        System.out.printf("Default cost: %.2f (freq floor=%.0f), Unknown cost: %.2f%n",
            defaultCost, MIN_FREQ_FLOOR, unknownCost);
    }

    /**
     * Build trie from dictionary for fast codepoint-based lookups.
     */
    private void buildTrie() {
        for (String word : words) {
            float cost = getWordCost(word);
            insertIntoTrie(word, cost);
        }
    }

    private void insertIntoTrie(String word, float cost) {
        TrieNode node = trie;
        int[] cps = word.codePoints().toArray();
        for (int cp : cps) {
            node = node.getOrCreateChild(cp);
        }
        node.isWord = true;
        node.cost = cost;
    }

    /**
     * Lookup codepoints in trie. Returns cost if found, null otherwise.
     */
    public Float lookupCodepoints(int[] cps, int start, int end) {
        TrieNode node = trie;
        for (int i = start; i < end; i++) {
            node = node.getChild(cps[i]);
            if (node == null) return null;
        }
        return node.isWord ? node.cost : null;
    }

    /**
     * Simple JSON parser for Map<String, Number> format.
     */
    private Map<String, Double> parseJsonMap(String json) {
        Map<String, Double> result = new HashMap<>();

        json = json.trim();
        if (json.startsWith("{")) json = json.substring(1);
        if (json.endsWith("}")) json = json.substring(0, json.length() - 1);

        StringBuilder keyBuilder = new StringBuilder();
        StringBuilder valueBuilder = new StringBuilder();
        boolean inKey = false;
        boolean inValue = false;
        boolean escaped = false;

        for (int i = 0; i < json.length(); i++) {
            char c = json.charAt(i);

            if (escaped) {
                if (inKey) keyBuilder.append(c);
                escaped = false;
                continue;
            }

            if (c == '\\') {
                escaped = true;
                continue;
            }

            if (c == '"') {
                if (!inKey && !inValue) {
                    inKey = true;
                    keyBuilder.setLength(0);
                } else if (inKey) {
                    inKey = false;
                }
                continue;
            }

            if (inKey) {
                keyBuilder.append(c);
                continue;
            }

            if (c == ':') {
                inValue = true;
                valueBuilder.setLength(0);
                continue;
            }

            if (c == ',' || c == '}') {
                if (inValue && valueBuilder.length() > 0) {
                    String key = keyBuilder.toString();
                    String valStr = valueBuilder.toString().trim();
                    try {
                        double val = Double.parseDouble(valStr);
                        result.put(key, val);
                    } catch (NumberFormatException e) {
                        // Skip invalid entries
                    }
                    inValue = false;
                }
                continue;
            }

            if (inValue) {
                valueBuilder.append(c);
            }
        }

        // Handle last entry
        if (inValue && valueBuilder.length() > 0) {
            String key = keyBuilder.toString();
            String valStr = valueBuilder.toString().trim();
            try {
                double val = Double.parseDouble(valStr);
                result.put(key, val);
            } catch (NumberFormatException e) {
                // Skip
            }
        }

        return result;
    }

    public boolean contains(String word) {
        return words.contains(word);
    }

    public float getWordCost(String word) {
        Float cost = wordCosts.get(word);
        if (cost != null) return cost;
        if (words.contains(word)) return defaultCost;
        return unknownCost;
    }

    /**
     * Trie node for efficient codepoint-based dictionary lookup.
     */
    private static class TrieNode {
        private Map<Integer, TrieNode> children;
        boolean isWord = false;
        float cost = 0;

        TrieNode getChild(int codepoint) {
            return children == null ? null : children.get(codepoint);
        }

        TrieNode getOrCreateChild(int codepoint) {
            if (children == null) {
                children = new HashMap<>();
            }
            return children.computeIfAbsent(codepoint, k -> new TrieNode());
        }
    }
}
